### PR TITLE
bpf: run kernel's checkpatch.pl locally and as GitHub action, fix style

### PR DESCRIPTION
**Pull request by qmonnet**
**BACKPORTS**: We want to backport the bpf/ changes (first 5 patches) to avoid conflicts with future backports. However it may not be suitable to backport the new GH action, so **last commit** (`build: run checkpatch.pl for bpf/ locally or as GitHub action`) **should be left out**.

---

Follow-up from RFC #11033.

Although there is no explicit coding style for the BPF code, several developers tend to follow [the kernel coding style conventions](https://www.kernel.org/doc/html/v5.6/process/coding-style.html). To try and maintain a consistent style all over the code, this PR adds kernel's checkpatch.pl to the list of checks performed on the source as part as BPF's `make check`.

Some details on the check:

- We only run checkpatch.pl on the bpf/ directory.

- A new `checkpatch` BPF make target is created, and added as a dependency to `check`.

- However, we only run it on the latest changes by default, meaning the changes since newest Git parent reference plus the changes returned by `git diff HEAD`.

- For manual checks, the bash script used to launch checkpatch has an option to run the checks on all the bpf/ source code, independently of the Git history.

- GitHub action runs checkpatch.pl on all commits from the PR. There is no mechanism to bypass checkpatch errors, Janitors will have to decide if issues reported by checkpatch should be ignored.

The checkpatch.pl and spelling.txt files (both under GPLv2) are directly copied from the kernel repository, as per commit f436a58e2619 (`"checkpatch: fix CONST_STRUCT when const_structs.checkpatch is missing"`) (in linux-next as of the creation of this patch). File `deprecated_terms.txt` is created from this PR, and populated in prevision of the changes proposed for the terminology. The script `checkpatch.sh` (bash script) launches checkpatch.pl with the relevant arguments and options. `COPYING` file is copied from `bpf/COPYING`, to make sure people get a version of the license boilerplate with the checkpatch code.

---

The PR also contains commits to fix a number of style errors and warnings reported by checkpatch.pl. See individual commit logs for details.

I believe fixing them improve code style consistency and readability, but it may complicate backports (although conflicts should be trivial to fix), so as always feedback is welcome.

### CODE REVIEW

**Comment by qmonnet** [comment]
test-me-please

File: .github/workflows/bpf-checks.yaml [hunk 1]
```diff
@@ -7,6 +7,17 @@ on:
...
 
 jobs:
+    checkpatch:
+        name: checkpatch
+        runs-on: ubuntu-latest
+        steps:
+            - name: Checkout code
+              uses: actions/checkout@v2
+              with:
+                  fetch-depth: 0
```
Comment (pchaigno) [inline .github/workflows/bpf-checks.yaml:17]: Did you find this makes a sensible difference? Should we use it in other GitHub workflows (not in this PR of course)?

File: contrib/checkpatch/checkpatch.pl [hunk 2]
```diff
@@ -0,0 +1,6937 @@
...
+my $help = 0;
+my $configuration_file = ".checkpatch.conf";
+my $max_line_length = 100;
+my $ignore_perl_version = 0;
+my $minimum_perl_version = 5.10.0;
+my $min_conf_desc_length = 4;
+my $spelling_file = "$D/spelling.txt";
+my $codespell = 0;
+my $codespellfile = "/usr/share/codespell/dictionary.txt";
+my $conststructsfile = "$D/const_structs.checkpatch";
```
Comment (pchaigno) [inline contrib/checkpatch/checkpatch.pl:64]: Couldn't we remove references to this file in the script instead of having an empty version? Unless you plan to keep this script up-to-date with the kernel's and want to avoid diverging too much? If that's the case, maybe it would be easier to download it directly from the kernel sources in the GitHub action.

**Review by pchaigno — COMMENTED** [review]
I'm a bit ashamed of asking this after you modified all files, but is there any benefit in applying all these style changes now? Couldn't we just way for peeps to modify the faulty lines, at which point `checkpatch.pl` will complain and the lines will be fixed? That would be a lot slower, but we wouldn't make backports any bit more difficult (nor would we break `git blame` :grimacing: ).

**Comment by qmonnet** [comment]
> is there any benefit in applying all these style changes now? Couldn't we just way for peeps to modify the faulty lines, at which point `checkpatch.pl` will complain and the lines will be fixed?

I know it's an intrusive change regarding backports/git history. My opinion is that code will look cleaner and more consistent so a bit easier to read/maintain. Keep in mind that most changes are chunks of only one or two lines, so it should be rather discreet in a `git blame`. Also it's very small/simple changes, so there may be conflicts for backports, but they should be trivial to solve.

This being said, if the consensus is that we don't want those changes, I can also drop these two patches and we can wait for the lines to be touched again to have the style fixed.

File: contrib/checkpatch/checkpatch.pl (hunk 2 above)
Comment (qmonnet) [reply contrib/checkpatch/checkpatch.pl:64]: Updating the script from the kernel repository from time to time was my intention indeed, so I'd rather avoid patching it if we don't have a strong incentive to do so. Having one empty file under contrib/checkpatch does not seem too expensive?

I thought of downloading the script from the kernel version, but I'd rather keep control of the version we use to avoid bad surprises (e.g. some change regarding a new file needed, that breaks checkpatch if the file is not found and makes all runs of the GitHub action fail all of a sudden.

Regarding the empty file, instead of patching our checkpatch version, I was considering maybe patching upstream so that it doesn't complain if the file is missing but `--ignore CONST_STRUCT` is passed. What do you think? (This specific check is also broken if `CONST_STRUCT` is not ignored but the file is present and empty by the way, so even if we don't really care in our case I could fix that too).

File: .github/workflows/bpf-checks.yaml (hunk 1 above)
Comment (qmonnet) [reply .github/workflows/bpf-checks.yaml:17]: Lower perf when checking out, but required because I run checkpatch on each individual commit from the PR. From what I understand, without it, checking out the repo is done in such a way it contains only one commit. Other GitHub actions as far as I can tell don't work on the individual commits so no need to change.

File: contrib/checkpatch/checkpatch.pl (hunk 2 above)
Comment (pchaigno) [reply contrib/checkpatch/checkpatch.pl:64]: > Regarding the empty file, instead of patching our checkpatch version, I was considering maybe patching upstream so that it doesn't complain if the file is missing but `--ignore CONST_STRUCT` is passed. What do you think? (This specific check is also broken if `CONST_STRUCT` is not ignored but the file is present and empty by the way, so even if we don't really care in our case I could fix that too).

:+1: if you have the time; otherwise LGTM!

Comment (qmonnet) [reply contrib/checkpatch/checkpatch.pl:64]: It doesn't look too hard (even though it's Perl!). I'll post that soon.

<details><summary>patch</summary>

``` diff
diff --git a/scripts/checkpatch.pl b/scripts/checkpatch.pl
index bf9e0e87a6ef..d91332360494 100755
--- a/scripts/checkpatch.pl
+++ b/scripts/checkpatch.pl
@@ -765,8 +765,10 @@ sub read_words {
 }
 
 my $const_structs = "";
-read_words(\$const_structs, $conststructsfile)
-    or warn "No structs that should be const will be found - file '$conststructsfile': $!\n";
+if (show_type("CONST_STRUCT")) {
+       read_words(\$const_structs, $conststructsfile)
+           or warn "No structs that should be const will be found - file '$conststructsfile': $!\n";
+}
 
 my $typeOtherTypedefs = "";
 if (length($typedefsfile)) {
@@ -6580,7 +6582,8 @@ sub process {
 
 # check for various structs that are normally const (ops, kgdb, device_tree)
 # and avoid what seem like struct definitions 'struct foo {'
-               if ($line !~ /\bconst\b/ &&
+               if ($const_structs ne "" &&
+                   $line !~ /\bconst\b/ &&
                    $line =~ /\bstruct\s+($const_structs)\b(?!\s*\{)/) {
                        WARN("CONST_STRUCT",
                             "struct $1 should normally be const\n" . $herecurr);
```
</details>

[Edit] Now submitted (https://lkml.kernel.org/patchwork/patch/1261008/).
v2: https://lkml.kernel.org/patchwork/patch/1261548/
v3 :astonished:: https://lkml.kernel.org/patchwork/patch/1263025/

File: bpf/lib/lb.h [hunk 3]
```diff
@@ -1266,18 +1260,15 @@ static __always_inline int lb4_local(const void *map, struct __ctx_buff *ctx,
...
-		}
 		slave = lb4_select_slave(svc->count);
 		slave_svc = lb4_lookup_slave(ctx, key, slave);
-		if (!slave_svc) {
+		if (!slave_svc)
 			goto drop_no_service;
-		}
 		backend = lb4_lookup_backend(ctx, slave_svc->backend_id);
-		if (backend == NULL) {
+		if (backend)
```
Comment (qmonnet) [inline bpf/lib/lb.h]: \*oopsies*

**Comment by qmonnet** [comment]
test-me-please

**Comment by qmonnet** [comment]
retest-net-next

**Comment by borkmann** [comment]
> > is there any benefit in applying all these style changes now? Couldn't we just way for peeps to modify the faulty lines, at which point `checkpatch.pl` will complain and the lines will be fixed?
> 
> I know it's an intrusive change regarding backports/git history. My opinion is that code will look cleaner and more consistent so a bit easier to read/maintain. Keep in mind that most changes are chunks of only one or two lines, so it should be rather discreet in a `git blame`. Also it's very small/simple changes, so there may be conflicts for backports, but they should be trivial to solve.
> 
> This being said, if the consensus is that we don't want those changes, I can also drop these two patches and we can wait for the lines to be touched again to have the style fixed.

I would vote for getting the BPF C code changes merged since i) Quentin already did the work and the code looks better indeed, and ii) it's not like we didn't do intrusive changes to the entire code case before in 1.8. Perhaps it even makes sense to get these backported for 1.8.1 to avoid future merge conflicts. I'd also like if we fix the BLOCK_COMMENT_STYLE one at some point to get consistency in there, but can be follow-up.

**Comment by qmonnet** [comment]
test-me-please

**Comment by qmonnet** [comment]
No problem for fixing `BLOCK_COMMENT_STYLE`, but yeah, I'll keep it for a follow-up PR. Sounds good for backports too, I can help with it if necessary.

**Comment by qmonnet** [comment]
test-me-please

**Comment by qmonnet** [comment]
Rebased on top of complexity changes, plus improved bash script and fixed additional warnings (see the two new patches).

<details><summary>Incremental diff for <code>checkpatch.sh</code></summary>

``` diff
diff --git a/contrib/checkpatch/checkpatch.sh b/contrib/checkpatch/checkpatch.sh
index a0d662d5f347..121917b51626 100755
--- a/contrib/checkpatch/checkpatch.sh
+++ b/contrib/checkpatch/checkpatch.sh
@@ -1,24 +1,42 @@
 #!/bin/bash
 
+# Default options for checkpatch
 options="--no-tree --strict --no-summary --show-types --color=always"
-ignores=" \
-    --ignore COMPLEX_MACRO \
-    --ignore MULTISTATEMENT_MACRO_USE_DO_WHILE \
-    \
-    --ignore BLOCK_COMMENT_STYLE \
-    --ignore CONSTANT_CONVERSION \
-    --ignore CONST_STRUCT \
-    --ignore EMAIL_SUBJECT \
-    --ignore JIFFIES_COMPARISON \
-    --ignore LEADING_SPACE \
-    --ignore MACRO_WITH_FLOW_CONTROL \
-    --ignore PRINTK_WITHOUT_KERN_LEVEL \
-    --ignore TRAILING_SEMICOLON \
-    --ignore TRAILING_STATEMENTS \
-    --ignore VOLATILE \
-    \
-    --ignore LONG_LINE_COMMENT \
-"
+
+# Report types to ignore
+ignore_list=(
+    # Errors
+    COMPLEX_MACRO
+    MULTISTATEMENT_MACRO_USE_DO_WHILE
+    # Warnings
+    CONSTANT_CONVERSION
+    CONST_STRUCT
+    EMAIL_SUBJECT
+    JIFFIES_COMPARISON
+    LEADING_SPACE
+    MACRO_WITH_FLOW_CONTROL
+    PRINTK_WITHOUT_KERN_LEVEL
+    TRAILING_SEMICOLON
+    TRAILING_STATEMENTS
+    VOLATILE
+    # Checks
+    LONG_LINE_COMMENT
+)
+ignores="--ignore $(IFS=,; echo "${ignore_list[*]}")"
+
+# Report types that checkpatch downgrades from warning to checks for --file
+type_list=(
+    AVOID_BUG
+    FSF_MAILING_ADDRESS
+    LONG_LINE
+    LONG_LINE_COMMENT
+    #LONG_LINE_STRING   # Not desired
+    #PREFER_FALLTHROUGH # fallthrough; not implemented
+    SPDX_LICENSE_TAG
+    TYPO_SPELLING
+)
+types="--types $(IFS=,; echo "${type_list[*]}")"
+
 script_dir="$(dirname $(realpath $0))"
 checkpatch="$script_dir/checkpatch.pl"
 
@@ -48,6 +66,10 @@ check_cmd() {
 
 update_sources() {
     sources=$(find bpf -name "*.[ch]" ! -path "bpf/include/elf/*" ! -path "bpf/include/linux/*")
+    if [ -z "$sources" ]; then
+        echo "Please run this script from the root of Cilium's repository."
+        exit 1
+    fi
 }
 
 check_commit() {
@@ -61,6 +83,7 @@ check_commit() {
 }
 
 all_code=0
+indulgent=0
 OPTIND=1
 while getopts "haiq" opt; do
     case "$opt" in
@@ -71,6 +94,7 @@ while getopts "haiq" opt; do
         all_code=1
         ;;
     i)
+        indulgent=1
         options="${options/--strict /}"
         ;;
     q)
@@ -84,9 +108,17 @@ shift $((OPTIND-1))
 # If -a option provided, simply run checkpatch on all *.c *.h code and exit
 if [ $all_code -eq 1 ]; then
     update_sources
+    echo -e "${HL_START}Checking files:$HL_END $(echo $sources | tr '\n' ' ')"
     ret=0
     $checkpatch $options $ignores -f $sources || ret=1
-    exit ret
+    if [ $indulgent -eq 1 ]; then
+        echo -e "${HL_START}Second run, to report 'checks' that should normally be 'warnings'...$HL_END"
+        # Re-run to cover types downgraded to checks by checkpatch when running
+        # on files, to be on par with what we do for commits.
+        $checkpatch $options --strict $types -f $sources || ret=1
+    fi
+    echo -e "${HL_START}All done$HL_END"
+    exit $ret
 fi
 
 check_cmd git jq
@@ -103,7 +135,7 @@ if [ -n "$GITHUB_REF" ]; then
 else
     # Running locally
     # We'll run checkpatch on each commit since newest parent ref
-    parent_ref=$(git log --simplify-by-decoration --pretty=format:'%D' -n 2 | sed -n '2{s/,.*//;p}')
+    parent_ref=$(git log --simplify-by-decoration --pretty=format:'%D' -n 2 | sed -n '2{s/,.*//;s/^tag: //;p}')
     list_commits=$(git log --pretty=format:"%H %s" $parent_ref.. | awk '
         BEGIN {print "["}
         {
```
</details>

**Review by aanm — APPROVED** [review]
Shouldn't the GH action pass?

**Comment by qmonnet** [comment]
> Shouldn't the GH action pass?

Checkpatch is not perfect and may raise false positives. In the case of this PR, we have:
- Some patches that touch lines without fixing all their format issues because some are fixed in a following patch (For example first patch fixes errors but leaves out warnings, which are fixed in the second patch). Because checkpatch.pl is run on each individual commits, it complains about those remaining warning-level issues in the first patch.
- Some patches touch lines that have other format issues I didn't want to fix in this series (e.g. camel case on `dstID`, should ideally be renamed `dst_id` or something like that), or that we don't want to fix at all (false positive with `MACRO_ARG_PRECEDENCE`, or `SPACING` errors where fields are aligned on several lines on purpose for example).

Short version: No, the GH action is not expected to pass for this PR. Plus it allows reviewers to see what the output is like :).

[Marking as depending on upstream because I sent a patch upstream to avoid having to create this empty const_structs.checkpatch file.]

**Comment by qmonnet** [comment]
So my patch for checkpatch got [merged](https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/commit/?id=f436a58e2619d7236916ca2c7f2818163826f1a4) into linux-next, so we don't have to keep an empty `const_struct.checkpatch` file anymore. Well in the meantime they added another file for “deprecated terms” (i.e. transitioning from loaded terminology), but I think that's a good opportunity to prevent re-introducing loaded terms that we are currently trying to remove.

The latest push has the updated checkpatch.pl from linux-next, and a short list of deprecated terms. Updates to bpf/ code have been rebased.

The PR is ready for review. I do not expect the related GH action to pass on this PR, for the reasons mentioned in my previous comment.

**Comment by qmonnet** [comment]
test-me-please

**Comment by qmonnet** [comment]
test-me-please

File: bpf/node_config.h [hunk 4]
```diff
@@ -16,20 +16,23 @@
...
 #define HOST_ID 1
 #define WORLD_ID 2
 #define UNMANAGED_ID 3
 #define HEALTH_ID 4
 #define INIT_ID 5
 #define REMOTE_NODE_ID 6
 #define HOST_IFINDEX_MAC { .addr = { 0xce, 0x72, 0xa7, 0x03, 0x88, 0x56 } }
-#define NAT46_PREFIX { .addr = { 0xbe, 0xef, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xa, 0x0, 0x0, 0x0, 0x0, 0x0 } }
+#define NAT46_PREFIX { .addr = { 0xbe, 0xef, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, \
+				 0x0, 0x0, 0xa, 0x0, 0x0, 0x0, 0x0, 0x0 } }
```
Comment (joestringer) [inline bpf/node_config.h]: Short line requirement is annoying here, makes the IP addresses less readable IMO :/

**Review by joestringer — APPROVED** [review]
Not sure if we want to leave a note somewhere in the tree where `checkpatch` came from / how to update it but LGTM.

**Review by borkmann — APPROVED** [review]
lgtm

**Comment by borkmann** [comment]
> No problem for fixing `BLOCK_COMMENT_STYLE`. Sounds good for backports too, I can help with it if necessary.

Yeah, please do, so we can also integrate for 1.8.1. Thanks!

File: bpf/node_config.h (hunk 4 above)
Comment (qmonnet) [reply bpf/node_config.h]: Right, ok I'll revert those. I see some conflicts appeared anyway.

**Comment by qmonnet** [comment]
>> No problem for fixing BLOCK_COMMENT_STYLE. Sounds good for backports too, I can help with it if necessary.
>
> Yeah, please do, so we can also integrate for 1.8.1. Thanks!

@borkmann I did it since we spoke of it, see [relevant commit in the PR](https://github.com/cilium/cilium/pull/11936/commits/dc9f5425a94f1aabd2be8b8c2dfb7d9430e13577). With the minor limitation that I couldn't find a way to enforce consistent marking on start of blocks (so both `net/`-style and `rest-of-kernel`-style are valid in the eyes of checkpatch, see details in commit log).

**Comment by borkmann** [comment]
> > > No problem for fixing BLOCK_COMMENT_STYLE. Sounds good for backports too, I can help with it if necessary.
> > 
> > 
> > Yeah, please do, so we can also integrate for 1.8.1. Thanks!
> 
> @borkmann I did it since we spoke of it, see [relevant commit in the PR](https://github.com/cilium/cilium/pull/11936/commits/dc9f5425a94f1aabd2be8b8c2dfb7d9430e13577). With the minor limitation that I couldn't find a way to enforce consistent marking on start of blocks (so both `net/`-style and `rest-of-kernel`-style are valid in the eyes of checkpatch, see details in commit log).

Good point, I looked only at 'Files changed', not the individual commits. What I meant above was whether checkpatch could also warn on C++ style comments (starting with `//`) so we can convert them over to block style and make it uniform?

**Comment by errordeveloper** [comment]
@qmonnet thanks for doing this! I wonder if the perl script itself needs to be checked in the repo at all, perhaps it could be a one of [`image-tools`](https://github.com/cilium/image-tools) container images?

**Comment by qmonnet** [comment]
> I wonder if the perl script itself needs to be checked in the repo at all, perhaps it could be packaged as a container image in [`cilium/image-tools`](https://github.com/cilium/image-tools)?

It could, but I'm not sure this is the best thing to do. The script might get some “breaking” changes (such as trying to read a new file, like it just did for the “deprecated terms”: If the file is not found it errors out, which would make the GH action fail). So I'd rather keep the updates manual so we can check there is no bad surprise. Plus it's about formatting anyway, so not a big deal if we don't get the “latest features” from the script.

**Comment by errordeveloper** [comment]
>It could, but I'm not sure this is the best thing to do. The script might get some “breaking” changes (such as trying to read a new file, like it just did for the “deprecated terms”: If the file is not found it errors out, which would make the GH action fail). So I'd rather keep the updates manual so we can check there is no bad surprise. Plus it's about formatting anyway, so not a big deal if we don't get the “latest features” from the script.

@qmonnet I wasn't suggesting that, I was suggesting that instead of having the script checked in, it'd be better to package it inside an image, and no need to pollute this repo with a huge script that is not written in any of the language that are common to the project and is under GPL (not APL).

So instead of checking the script it here, it would a concern of the image-tools repo to either download on image build, or have it checked in (if really needed). But it certain won't be downloaded at runtime.

And in this repo (and possibly other repos), it could be consumed like this:
```
      - uses: actions/checkout@v1
      - uses: docker://docker.io/cilium/checkpatch:3e2ea4f151593908c362307a1de22e68610d955c
        name: Run checkpatch.pl
```
And in a makefile:
```
checkpatch:
	@$(ECHO_CHECK) "(checkpatch)"
	$(QUIET) docker run <args> docker.io/cilium/checkpatch:3e2ea4f151593908c362307a1de22e68610d955c
```

So instead of being part of a source tree, it would be a packaged tool basically.

**Comment by qmonnet** [comment]
@errordeveloper Oh, I had missed your point indeed. Hmm. Yeah, I think this is a good idea. Thanks!

Given that the current version is more or less ready and that the PR has to be rebased often, I will implement your suggestion in a follow-up, if you don't mind.

**Comment by qmonnet** [comment]
PR rebased on top of @pchaigno's changes.

> Good point, I looked only at 'Files changed', not the individual commits. What I meant above was whether checkpatch could also warn on C++ style comments (starting with `//`) so we can convert them over to block style and make it uniform?

@borkmann Right, I missed those. This is because checkpatch tolerates them by default, you have to pass it an `--ignore` option to actually _activate_ those reports. Such a good idea, oh well. So anyway, **I now added a commit to deal with `// ` comments and turn them into regular `/* ... */` blocks.** One drawback: checkpatch.pl also reports the `// SPDX-License-Identifer` tags in .c files as infringing the rule (but I left them untouched). Probably worth another patch upstream when I have some time.

> Not sure if we want to leave a note somewhere in the tree where `checkpatch` came from / how to update it

@joestringer Good idea, I left a comment with the URL to checkpatch.pl (from git repo at kernel.org) in the bash script.

**Comment by qmonnet** [comment]
test-me-please

**Comment by qmonnet** [comment]
retest-net-next

**Comment by errordeveloper** [comment]
>  I will implement your suggestion in a follow-up, if you don't mind.

@qmonnet of course! Hit me up when you find the time to do it :)

**Comment by borkmann** [comment]
> PR rebased on top of @pchaigno's changes.
> 
> > Good point, I looked only at 'Files changed', not the individual commits. What I meant above was whether checkpatch could also warn on C++ style comments (starting with `//`) so we can convert them over to block style and make it uniform?
> 
> @borkmann Right, I missed those. This is because checkpatch tolerates them by default, you have to pass it an `--ignore` option to actually _activate_ those reports. Such a good idea, oh well. So anyway, **I now added a commit to deal with `// ` comments and turn them into regular `/* ... */` blocks.** One drawback: checkpatch.pl also reports the `// SPDX-License-Identifer` tags in .c files as infringing the rule (but I left them untouched). Probably worth [another patch upstream](https://lore.kernel.org/patchwork/patch/1265784/) when I have some time.

Cool, thanks for fixing these up as well! Yeah SPDX need to be left untouched. Probably indeed worth to fix upstream so we have no errors on them.

**Comment by qmonnet** [comment]
> Cool, thanks for fixing these up as well! Yeah SPDX need to be left untouched. Probably indeed worth to fix upstream so we have no errors on them.

@borkmann I sent a patch but it will probably not make it, Joe P. doesn't think this is a useful addition :( https://lore.kernel.org/patchwork/patch/1265784/#1461076
